### ISSUE TITLE
Temporal difference for each episode is not captured

### ISSUE THREAD

**Issue by UDynamic** [UDynamic/Blackjack-Q-learning-AI-Agent#2]
the env wrapper only saves last 100 episodes data,

implemented a solution for capturing reward and steps length for each episode but still haven't made a solution for temporal difference

**Comment by UDynamic** [comment]
<html>
<body>
<html><head></head><body><p>Excellent — you’re absolutely right to think about tracking <strong>temporal difference (TD) errors</strong> for all episodes, not just the last few. Let’s go step by step.</p>
<hr>
<h2>🧠 What’s Happening Now</h2>
<p><code inline="">RecordEpisodeStatistics</code> stores <strong>per-episode cumulative reward</strong> and <strong>episode length</strong> in <strong>two deques</strong>:</p>
<ul>
<li>
<p><code inline="">env.return_queue</code> → total rewards per episode</p>
</li>
<li>
<p><code inline="">env.length_queue</code> → lengths per episode</p>
</li>
</ul>
<p>By default, both are limited to <strong>the most recent 100 episodes</strong> (<code inline="">deque_size=100</code>).<br>
So yes — you lose older history.</p>
<hr>
<h2>⚙️ Why This Limitation Exists</h2>
<p>The <code inline="">deque_size=100</code> default is meant for <strong>real-time monitoring</strong> (so memory doesn’t grow infinitely during long training).<br>
For small experiments, that’s fine — but for <strong>analysis or plotting full training curves</strong>, it’s too small.</p>
<hr>
<h2>✅ Solutions</h2>
<h3>🩵 Option 1: Record Everything Yourself</h3>
<p>The cleanest approach is to create your own Python lists to log <strong>per-episode rewards, lengths, and TD errors</strong> — independent of the <code inline="">RecordEpisodeStatistics</code> wrapper.</p>
<p>Here’s how you can modify your training loop:</p>
<pre><code class="language-python">episode_rewards = []
episode_lengths = []
episode_td_errors = []  # NEW

for episode in tqdm(range(n_episodes)):
    obs, info = env.reset()
    done = False
    total_reward = 0
    total_td_error = 0
    steps = 0

    while not done:
        action = agent.get_action(obs)
        next_obs, reward, terminated, truncated, info = env.step(action)

        # Update agent and get TD error
        td_error = agent.update(obs, action, reward, terminated, next_obs)
        total_td_error += abs(td_error)  # track mean magnitude
        total_reward += reward
        steps += 1

        done = terminated or truncated
        obs = next_obs

    episode_rewards.append(total_reward)
    episode_lengths.append(steps)
    episode_td_errors.append(total_td_error / steps)  # average TD error per episode

    agent.decay_epsilon()
</code></pre>
<p>You’ll just need to modify your agent’s <code inline="">update()</code> method to <strong>return</strong> the TD error each time:</p>
<pre><code class="language-python">def update(self, state, action, reward, done, next_state):
    td_target = reward + self.gamma * np.max(self.Q[next_state]) * (1 - done)
    td_error = td_target - self.Q[state][action]
    self.Q[state][action] += self.lr * td_error
    self.training_error.append(abs(td_error))  # for running avg
    return td_error  # &lt;-- new return value
</code></pre>
<hr>
<h3>🩷 Option 2: Increase <code inline="">deque_size</code></h3>
<p>If you still want to rely on <code inline="">RecordEpisodeStatistics</code>, you can re-create it with a <strong>larger buffer</strong> (say, <code inline="">deque_size=10000</code>), provided your memory allows it:</p>
<pre><code class="language-python">env = gym.wrappers.RecordEpisodeStatistics(env, deque_size=10000)
</code></pre>
<p>That will store up to the last 10,000 episodes — but not <em>all</em> if you exceed that.</p>
<hr>
<h2>🧩 Visualizing Everything Later</h2>
<p>After training, you can then plot:</p>
<pre><code class="language-python">plt.figure(figsize=(10, 5))
plt.plot(episode_rewards, label="Rewards")
plt.plot(episode_td_errors, label="TD Error")
plt.legend()
plt.title("Training Performance")
plt.show()
</code></pre>
<hr>
<h2>🪄 Best Practice Summary</h2>

Goal | Method
-- | --
Track full training stats | Use your own lists (episode_rewards, episode_td_errors, etc.)
Keep memory low | Use RecordEpisodeStatistics with small deque_size
Want both | Combine both methods — wrapper for live stats, lists for full logs
Want smooth visualization | Apply rolling average (np.convolve) on your lists


<hr>
<p>Would you like me to show you how to integrate <strong>both logging and visualization of TD error + reward curves</strong> into your full training script (cleanly structured)?</p></body></html>
</body>
</html>
